golang/cosmos/x/vibc/types/receiver.go

```diff
@@ -0,0 +1,177 @@
+	ReceiveChanCloseInit(ctx sdk.Context, sourcePort, sourceChannel string) error
+	ReceiveBindPort(ctx sdk.Context, sourcePort string) error
+	ReceiveTimeoutExecuted(ctx sdk.Context, packet exported.PacketI) error
+}
+
+type Receiver struct {
+	impl ReceiverImpl
+}
+
+func NewReceiver(impl ReceiverImpl) Receiver {
+	return Receiver{
+		impl: impl,
+	}
+}
+
+type portMessage struct { // comes from swingset's IBC handler
+	Type              string              `json:"type"` // IBC_METHOD
+	Method            string              `json:"method"`
+	Packet            channeltypes.Packet `json:"packet"`
+	RelativeTimeoutNs uint64              `json:"relativeTimeoutNs,string"`
+	Order             string              `json:"order"`
+	Hops              []string            `json:"hops"`
+	Version           string              `json:"version"`
+	Ack               []byte              `json:"ack"`
+}
+
+func stringToOrder(order string) channeltypes.Order {
+	switch order {
+	case "ORDERED":
+		return channeltypes.ORDERED
+	case "UNORDERED":
+		return channeltypes.UNORDERED
+	default:
+		return channeltypes.NONE
+	}
+}
+
+func orderToString(order channeltypes.Order) string {
+	switch order {
+	case channeltypes.ORDERED:
+		return "ORDERED"
+	case channeltypes.UNORDERED:
+		return "UNORDERED"
+	default:
+		return "NONE"
+	}
+}
+
+type rawAcknowledgement struct {
+	data []byte
+}
+
+func (r rawAcknowledgement) Acknowledgement() []byte {
+	return r.data
+}
+
+func (r rawAcknowledgement) Success() bool {
+	return true
+}
+
+func (ir Receiver) Receive(cctx context.Context, str string) (ret string, err error) {
+	// fmt.Println("ibc.go downcall", str)
+	ctx := sdk.UnwrapSDKContext(cctx)
+	impl := ir.impl
+
+	msg := new(portMessage)
+	err = json.Unmarshal([]byte(str), &msg)
+	if err != nil {
+		return ret, err
+	}
+
+	if msg.Type != "IBC_METHOD" {
+		return "", fmt.Errorf(`channel handler only accepts messages of "type": "IBC_METHOD"`)
+	}
+
+	switch msg.Method {
+	case "sendPacket":
+		timeoutTimestamp := msg.Packet.TimeoutTimestamp
+		if msg.Packet.TimeoutHeight.IsZero() && timeoutTimestamp == 0 {
+			// Use the relative timeout if no absolute timeout is specifiied.
+			timeoutTimestamp = uint64(ctx.BlockTime().UnixNano()) + msg.RelativeTimeoutNs
+		}
+
+		packet := channeltypes.NewPacket(
+			msg.Packet.Data, 0,
+			msg.Packet.SourcePort, msg.Packet.SourceChannel,
+			msg.Packet.DestinationPort, msg.Packet.DestinationChannel,
+			msg.Packet.TimeoutHeight, timeoutTimestamp,
+		)
+		seq, err := impl.ReceiveSendPacket(ctx, packet)
+		if err == nil {
+			packet.Sequence = seq
+			bytes, err := json.Marshal(&packet)
+			if err == nil {
+				ret = string(bytes)
+			}
+		}
+
+	case "tryOpenExecuted":
+		err = impl.ReceiveWriteOpenTryChannel(
+			ctx, msg.Packet,
+			stringToOrder(msg.Order), msg.Hops, msg.Version,
+		)
+		if err == nil {
+			ret = "true"
```

Comment:
Consider using a constant for `"true"` so you won't get tripped up on a misspelling.

Also consider setting it after the switch instead:

```
if ret == "" && err == nil {
        ret = "true" // or a defined constant
}
```